pca.go

```diff
@@ -0,0 +1,76 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix"
+	"github.com/gonum/matrix/mat64"
+)
+
+// PrincipalComponents returns the k principal component direction vectors of
+// the matrix a projected score vectors and variance values by singular value
+// decomposition of the input. The input a is a matrix where variables are
+// represented in the columns and observations are the rows in the input.
+//
+// If weights is not nil, it is used to weight the observations. The length of
+// weight must match the number of observations or PrincipalComponents will panic.
+//
+// If no singular value decomposition is possible, pc, proj and vars are returned
+// nil and ok is returned false. PrincipalComponents will panic if k is greater
+// than or equal to the number of rows, or if k is greater than the number of
+// columns.
+func PrincipalComponents(a mat64.Matrix, weights []float64, k int) (pc, proj *mat64.Dense, vars []float64, ok bool) {
+	n, d := a.Dims()
+	if k >= n {
+		panic("stat: k >= observations")
+	}
+	if k > d {
+		panic("stat: k > variables")
+	}
+	if weights != nil && len(weights) != n {
+		panic("stat: len(weights) != observations")
+	}
+
+	centered := mat64.NewDense(n, d, nil)
+	f := 1 / float64(n)
+	for j := 0; j < d; j++ {
```

Comment:
Which says "First, one needs to compute the weighted mean μ=1n∑wixi and subtract it from the data in order to center it."